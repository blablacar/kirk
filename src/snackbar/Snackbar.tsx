import React, { PureComponent } from 'react'
import { createPortal } from 'react-dom'
import TransitionGroup from 'react-transition-group/TransitionGroup'
import cc from 'classcat'
import { canUseDOM } from 'exenv'

import { color } from '_utils/branding'
import Button, { ButtonStatus } from 'button'
import CrossIcon from 'icon/crossIcon'
import Text, { TextDisplayType } from 'text'
import CustomTransition, { AnimationType } from 'transitions'

export interface SnackbarProps {
  readonly close: () => void
  readonly isOpen: boolean
  readonly className?: string
  readonly extraClassName?: string
  readonly children: JSX.Element
}

class Snackbar extends PureComponent<SnackbarProps> {
  private portalNode: HTMLElement
  static defaultProps: Partial<SnackbarProps> = {
    isOpen: false,
  }

  constructor(props: SnackbarProps) {
    super(props)
    if (canUseDOM) {
      this.portalNode = document.createElement('div')
      document.body.appendChild(this.portalNode)
    }
  }

  componentWillUnmount() {
    if (this.portalNode) {
      document.body.removeChild(this.portalNode)
      this.portalNode = null
    }
  }

  render() {
    const modalElement = (
      <TransitionGroup
        component="div"
        className={cc(['transition-wrapper kirk-snackbar-container', this.props.className])}
      >
        {this.props.isOpen && (
          <CustomTransition animationName={AnimationType.SLIDE_UP}>
            <div className={cc(['kirk-snackbar', this.props.extraClassName])}>
              <Text role="alert" display={TextDisplayType.TITLE} className="kirk-snackbar-content">
                {this.props.children}
              </Text>
              <Button
                status={ButtonStatus.UNSTYLED}
                className="kirk-snackbar-cross"
                onClick={this.props.close}
              >
                <CrossIcon iconColor={color.white} size="18px" />
              </Button>
            </div>
          </CustomTransition>
        )}
      </TransitionGroup>
    )

    if (!this.portalNode) {
      return modalElement
    }

    return createPortal(modalElement, this.portalNode)
  }
}

export default Snackbar
