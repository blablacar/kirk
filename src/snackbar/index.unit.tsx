import React from 'react'
import renderer from 'react-test-renderer'
import Snackbar from 'snackbar'
import Button from 'button'
import exenv from 'exenv'

const close = jest.fn()

describe('Snackbar', () => {
  it('should show the snackbar when isOpen is true', () => {
    const wrapper = shallow(<Snackbar isOpen close={close} />)
    expect(wrapper.find('.kirk-snackbar').exists()).toBe(true)
  })
  it('should not show the snackbar when isOpen is false', () => {
    const wrapper = shallow(<Snackbar isOpen={false} close={close} />)
    expect(wrapper.find('.kirk-snackbar').exists()).toBe(false)
  })
  it('should call close method when click on cross button', () => {
    const wrapper = shallow(<Snackbar isOpen close={close} />)
    wrapper.find(Button).simulate('click')
    expect(close).toHaveBeenCalled()
  })
  it('should render the same layout on client and server side', () => {
    exenv.canUseDOM = false
    const serverSide = mount(<Snackbar isOpen close={close} />)
    exenv.canUseDOM = true
    const clientSide = mount(<Snackbar isOpen close={close} />)

    expect(serverSide.html()).toEqual(clientSide.html())
  })
  it('should not have changed', () => {
    exenv.canUseDOM = false
    const snackbarServerSide = renderer
      .create(
        <Snackbar isOpen close={close}>
          Oups
        </Snackbar>,
      )
      .toJSON()
    exenv.canUseDOM = true
    expect(snackbarServerSide).toMatchSnapshot()
  })
})
