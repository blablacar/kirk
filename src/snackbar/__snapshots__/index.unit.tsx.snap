// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar should not have changed 1`] = `
<div
  className="transition-wrapper kirk-snackbar-container"
>
  <div
    className="jsx-2822186252 kirk-snackbar slide-up slide-up-entered"
  >
    <span
      className="kirk-text kirk-text-title kirk-snackbar-content"
      role="alert"
    >
      Oups
    </span>
    <button
      className="kirk-button kirk-button-unstyled kirk-snackbar-cross"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="kirk-icon"
        height="18px"
        viewBox="0 0 24 24"
        width="18px"
        xmlns="http://www.w3.org/2000/svg"
      >
        
        <path
          d="M19 5L5 19M19 19L5 5"
          fill="none"
          stroke="#FFF"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeMiterlimit="10"
          strokeWidth="2"
        />
      </svg>
    </button>
  </div>
</div>
`;
